### ISSUE TITLE
Safer access to HDF5 attributes

### ISSUE THREAD

**Issue by ajelenak** [h5py/h5py#2435]
I've been involved with two separate projects in the last two weeks that had the same problem: inadvertently recreating HDF5 dimension scale-specific attributes between datasets in two different files and, therefore, transferring invalid dimension scale information. It may be useful to have a "safer" access to HDF5 attributes that excludes the reserved ones. I don't have a specific implementation for this, just wanted to record the problem here for possible discussion. Adding a warning to the attributes documentation could be the first step. (But only for those who read documentation. 😄)

**Comment by takluyver** [comment]
Could you explain the problem you saw a bit more, I'm not quite getting it? Did someone inadvertantly use the dimension scale attribute names (`DIMENSION_LIST` etc.) for something else just by chance? Or did they copy all attributes from one dataset to another without checking if that made any sense?

**Comment by ajelenak** [comment]
> Or did they copy all attributes from one dataset to another without checking if that made any sense?

This.

**Comment by takluyver** [comment]
To my mind, there's nothing particularly special about HDF5 dimension scale attributes when copying. Attributes are metadata on a dataset, and they may or may not be applicable to another dataset. E.g. in a NetCDF4 file, a format built on top of HDF5, there's a `_Netcdf4Coordinates` variable, which is logically similar to `DIMENSION_LIST` (in fact, both are used together). Copying that to a dataset with different axes would be equally broken.

I notice also that H5Ocopy doesn't seem to treat dimension scale attributes specially - it doesn't update the other side of the DIMENSION_LIST / REFERENCE_LIST pair to maintain the two-way linking. (Aside: do you know why these links go both ways? It seems like a very brittle scheme.)

**Comment by ajelenak** [comment]
I agree this issue is close to 50-50 but I want to see if there is anything that can be done to protect less knowledgeable users. Putting a warning in documentation seems like the minimum.

The `_Netcdf4Coordinates` attribute is a good example but I consider it the domain of h5netcdf and Python packages that use libnetcdf. The `ncdump` tool does not show it when listing netCDF-4 file content but `h5dump` shows it. Is it possible to do something similar in h5py? Have a way to list only non-reserved attributes? This way the list of these attributes would be in one place.

> I notice also that H5Ocopy doesn't seem to treat dimension scale attributes specially - it doesn't update the other side of the DIMENSION_LIST / REFERENCE_LIST pair to maintain the two-way linking. (Aside: do you know why these links go both ways? It seems like a very brittle scheme.)

No, I don't know why dimension scales were implemented this way. Dimension scale API is considered high level to libhdf5 which I guess is a different way to say "someone else's problem". 😄

**Comment by epourmal** [comment]
> (Aside: do you know why these links go both ways? It seems like a very brittle scheme.)
Probably, I can shade some light. Dimension scales were implemented to address netCDF-4 requirements and to help HDF4 users/apps to move to HDF5.  The links allow to find all dimension scales associated with dataset's dimension AND find all datasets that share the same dimension.

**Comment by epourmal** [comment]
> notice also that H5Ocopy doesn't seem to treat dimension scale attributes specially - it doesn't update the other side of the DIMENSION_LIST / REFERENCE_LIST pair to maintain the two-way linking.

I think this is an oversight. One could argue that HDF5 shouldn't care about attributes created by high-level APIs but not in the case of HDF5 HL APIs. The function should be more flexible and update both lists if the corresponding datasets are copied during the same copy operation, but you can see why it could be tricky.
